warehouse/macaroons/caveats/_legacy.py

```diff
@@ -0,0 +1,79 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections.abc import Mapping, Sequence
+
+from pyramid.threadlocal import get_current_request
+
+
+def _adapt_v1(data: Mapping) -> Sequence | None:
+    permissions = data.get("permissions")
+
+    # This would be a malformed token, so we'll just refuse
+    # to adapt it.
+    if permissions is None:
+        return None
+
+    # Our V1 token didn't have a way to specify that a token should be
+    # restricted to a specific user, just that it was scoped to "the user",
+    # which the user was whoever the token was linked to in the database.
+    # Our new tokens strengthens that to validate that the linked user
+    # matches who it is expected to be, but since we don't have that
+    # data for V1 tokens, we'll just use the current user.
+    if permissions == "user":
```

Comment:
Just for visibility: https://github.com/pypi/warehouse/pull/11272 will add `"oidc"` here, but if we merge this first (and we should!) we can avoid having to adapt it 🙂

Reply:
Yea, I would expect OIDC to make something like:

```python
@as_caveat(tag=N)
@dataclass(frozen=True)
class RequestOIDC(Caveat):
    oidc_provider_id: StrictStr

    def verify(self, request: Request, context: Any, permission: str) -> Result:
        if not isinstance(request.identity, OIDCProvider):
            return Failure("token with oidc provider restriction without a OIDC Provider")

        if str(request.identity.id) != self.user_id:
            return Failure("current oidc provider does not match oidc provider restriction in token")

        return Success()
```

Which has the same (new) semantics, we're not pulling the oidc (or user) out of the caveats, but rather the token include an assertion that whatever ``request.identity`` is associated with the request has to be the expected OIDC Provider (or User).

The caveat itself doesn't care about *how* ``request.identity`` got set, it's just verifying it's the correct one. Kind of silly when the token identifier is how we're looking up that ``request.identity``, but it adds a slight protection against confusion attacks where the ``request.identity`` comes from a different authentication source than the macaroon itself [^1].

[^1]: In theory this shouldn't happen, but our code base doesn't *prevent* it, so it's possible?